lib/ddtrace/utils.rb

```diff
@@ -1,9 +1,21 @@
+require 'thread'
+
 module Datadog
-  # TODO[manu]: write docs
+  # Utils contains low-level utilities, typically to provide pseudo-random trace IDs.
   module Utils
+    # We use a custom random number generator because we want no interference
+    # with the default one. Using the default prng, we could break code that
+    # would rely on srand/rand sequences.
+    @rnd = Random.new
+    # We want to garantee all calls are sequential else, in some edge cases,
+    # the same value could be return twice, ending in trace collision.
```

Comment:
I should dig dipper but reading this https://github.com/ruby/ruby/blob/trunk/random.c#L183 I see no protection whatsoever from running this in parallel. You don't  necessarily need to lock to get a random number but if you want to re-use the same generator across different threads -> yes you need. And this is what you'd do by using the default `rand`. But (my bad) as this is a C impl, it should be protected by the GIL so the collision is truely impossible (unless the GIL is removed, but this is not going to happen soon, I hear it could be planned for Ruby 3, but no hurry). Will remove the lock & leave a comment.